enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,424 @@
+the condition will be marked false.
+
+### Risks and Mitigations
+
+* Annotation keys must conform to length limits: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set
+* Requires well-behaved controllers and admins to keep things running
+smoothly. Would be easy to disrupt machines with poor configuration.
+* Troubleshooting problems may increase in complexity, but this is
+mitigated mostly by the fact that these hooks are opt-in. Operators
+will or should know they are consuming these hooks, but a future proliferation
+of the cluster-api could result in these components being bundled as a
+complete solution that operators just consume. To this end, we should
+update any troubleshooting guides to check these hook points where possible.
+* Users may add hooks which end up indefinitely blocking removal of Machines.
+We already have a [MachineNotYetDeleted](https://github.com/openshift/machine-api-operator/blob/master/docs/user/Alerts.md#machinenotyetdeleted)
+alert that will warn users when a Machine removal is blocked for an extended
+period. This should continue to function and notify users when a removal has
+been blocked, at which point we expect manual intervention to check the Machine.
+
+## Design Details
+
+### Test Plan
+
+* E2E testing will be added to the cluster-api-actuator-pkg (Machine API E2E suite)
+to ensure that the hooks behave as expected
+
+### Graduation Criteria
+
+This feature is not large enough to require graduation and will be considered GA from
+initial release.
+
+#### Dev Preview -> Tech Preview
+
+This feature will be released straight to GA.
+
+#### Tech Preview -> GA
+
+This feature will be released straight to GA.
+
+#### Removing a deprecated feature
+
+This proposal does not involve removing any deprecated features.
+
+### Upgrade / Downgrade Strategy
+
+As this feature is opt-in, there is no need to account for upgrades.
+
+Any new Hook Implementing Controller will be able to add the hooks at
+any point and they will be observed when the updated Machine controller
+is deployed via the CVO.
+
+On downgrades, HICs may wish to remove the annotations, but they are not
+harmful if left behind.
+
+### Version Skew Strategy
+
+We do not expect any issues with version skew as this is a new feature.
+
+## Implementation History
+
+* This has already been implemented in Cluster API
+
+## Drawbacks
+
+* If misused, this could prevent Machines from being deleted and lead to an increase in support
+cases to determine if it is safe to override the feature.
+
+## Alternatives
+
+### Custom Machine Controller
+Require advanced users to fork and customize.
+This can already be done if someone chooses, so not much of a solution.
+
+### Finalizers
+We define additional finalizers, but this really only implies the deletion lifecycle point.
+A misbehaving controller that accidentally removes finalizers could have undesirable effects.
```

Comment:
doesn't the misbehaving bit also apply to the annotation?

Reply:
Updated this with a good example, if you remove too many finalizers, the Machine gets removed from etcd, possibly before the machine has been deleted on the cloud side, leaking the instance and adding cost. Removing too many finalizers wouldn't result in the machine instance being leaked.